module/zfs/zfs_ioctl.c

```diff
@@ -5232,25 +5499,992 @@ zfs_ioc_send_space(const char *snapname, nvlist_t *innvl, nvlist_t *outnvl)
+					dmu_flags |= DS_FIND_SNAPSHOTS;
+			}
+
+			err = dmu_objset_find_dp_impl(dp, dp->dp_root_dir_obj,
+			    &dump_list_strategy_cb, dls, dmu_flags,
+			    dls->dls_depth);
+			dsl_pool_config_exit(dp, FTAG);
+			dls->dls_flags = dls_flags;
+			if (err)
+				break;
+
+		}
+		mutex_exit(&spa_namespace_lock);
+	}
+
+	ASSERT(err < 256);
+	bzero(&zpr, sizeof (zfs_pipe_record_t));
+	zpr.zpr_err = (uint8_t) err;
+	(void) vn_rdwr(UIO_WRITE, dls->dls_vp, (caddr_t) &zpr,
+	    sizeof (uint64_t), 0, UIO_SYSSPACE, FAPPEND, RLIM64_INFINITY,
+	    CRED(), &resid);
+
+	areleasef(dls->dls_fd, dls->dls_fip);
+	kmem_free(dls->dls_fsname, sizeof (dls_t));
+	kmem_free(dls, sizeof (dls_t));
+}
+
+static int
+zfs_stable_ioc_zfs_list(const char *fsname, nvlist_t *innvl,
+    nvlist_t *outnvl, nvlist_t *opts, uint64_t version)
+{
+	dsl_pool_t *dp = NULL;
+	int fd;
+	file_t *fp;
+	nvlist_t *nv;
+	dls_t *dls;
+	dls_flag_t dls_flags = 0;
+	uint64_t depth = 0-1;
+	uint64_t dd_obj = 0;
+	int error;
+
+	error = nvlist_lookup_int32(innvl, "fd", &fd);
+	if (error != 0)
+		return (SET_ERROR(EINVAL));
+
+	if (nvlist_exists(opts, "recurse")) {
+		dls_flags |= DLS_RECURSE;
+		(void) nvlist_lookup_uint64(opts, "recurse", &depth);
+	}
+
+	if (nvlist_lookup_nvlist(opts, "type", &nv) == 0) {
+		if (nvlist_exists(nv, "all"))
+			dls_flags |= DLS_TRAVERSE_ALL;
+		else {
+			if (nvlist_exists(nv, "bookmark"))
+				dls_flags |= DLS_TRAVERSE_BOOKMARK;
+			if (nvlist_exists(nv, "filesystem"))
+				dls_flags |= DLS_TRAVERSE_FILESYSTEM;
+			if (nvlist_exists(nv, "snap"))
+				dls_flags |= DLS_TRAVERSE_SNAPSHOT;
+			if (nvlist_exists(nv, "snapshot"))
+				dls_flags |= DLS_TRAVERSE_SNAPSHOT;
+			if (nvlist_exists(nv, "volume"))
+				dls_flags |= DLS_TRAVERSE_VOLUME;
+		}
+	}
+
+	/* Pass an object number when given a DSL directory name */
+	if (fsname != NULL && fsname[0] != '\0') {
+		spa_t *spa;
+		dsl_dataset_t *ds = NULL;
+
+		mutex_enter(&spa_namespace_lock);
+
+		spa = spa_lookup(fsname);
+		if (spa == NULL) {
+			mutex_exit(&spa_namespace_lock);
+			return (ENOENT);
+		}
+
+		dp = spa_get_dsl(spa);
+
+		dsl_pool_config_enter(dp, FTAG);
+		mutex_exit(&spa_namespace_lock);
+
+		error = dsl_dataset_hold(dp, fsname, FTAG, &ds);
+
+		/* Adopt sane defaults based on the DSL directory */
+		if (!nvlist_exists(nv, "type")) {
+			if ((strchr(fsname, '#') != NULL)) {
+				dls_flags |= DLS_TRAVERSE_BOOKMARK;
+				dls_flags &= ~DLS_RECURSE;
+			} else if ((strchr(fsname, '@') != NULL)) {
+				dls_flags |= DLS_TRAVERSE_SNAPSHOT;
+				dls_flags &= ~DLS_RECURSE;
+			} else {
+				objset_t *osp;
+				dmu_objset_from_ds(ds, &osp);
+
+				switch (dmu_objset_type(osp)) {
+				case DMU_OST_ZVOL:
+					dls_flags |= DLS_TRAVERSE_VOLUME;
+					break;
+				case DMU_OST_ZFS:
+					dls_flags |= DLS_TRAVERSE_FILESYSTEM;
+					break;
+				default:
+					VERIFY(0);
+				}
+			}
+		}
+
+		dsl_pool_config_exit(dp, FTAG);
+		if (error)
+			return (error);
+		dd_obj = ds->ds_dir->dd_object;
```

Comment:
There is a race here between `zfs list` and `zfs destroy`. I am not sure what to do about it (if anything).